crypto/src/mls/context.rs

```diff
@@ -0,0 +1,167 @@
+
+use std::{ops::Deref, sync::Arc};
+
+use async_lock::{MutexGuardArc, RwLock, RwLockReadGuardArc, RwLockWriteGuardArc};
+use core_crypto_keystore::KeystoreTransaction;
+use mls_crypto_provider::TransactionalCryptoProvider;
+
+use crate::{
+    group_store::GroupStore,
+    prelude::{Client, MlsConversation},
+    CoreCryptoCallbacks, CryptoError, CryptoResult,
+};
+
+use super::MlsCentral;
+
+/// This struct provides transactional support for Core Crypto.
+///
+/// This is struct provides mutable access to the internals of Core Crypto. Every operation that
+/// causes data to be persisted needs to be done through this struct. This struct will buffer all
+/// operations in memory and when [CentraContext::finish] is called, it will persist the data into
+/// the keystore.
+#[derive(Debug, Clone)]
+pub struct CentralContext {
+    state: Arc<RwLock<ContextState>>,
+    _lock_guard: Arc<MutexGuardArc<()>>,
+}
+
+/// Due to uniffi's design, we can't force the context to be dropped after the transaction is
+/// committed. To work around that we switch the value to `Invalid` when the context is finished
+/// and throw errors if something is called
+#[derive(Debug, Clone)]
+enum ContextState {
+    Valid {
+        transaction: TransactionalCryptoProvider,
+        callbacks: Arc<RwLock<Option<std::sync::Arc<dyn CoreCryptoCallbacks + 'static>>>>,
+        mls_client: Arc<RwLock<Option<Client>>>,
+        mls_groups: Arc<RwLock<GroupStore<MlsConversation>>>,
+    },
+    Invalid,
+}
+
+impl MlsCentral {
+    /// Creates a new transaction within the MlsCentral. All operations that persist data will be
+    /// buffered in memory and when [CentralContext::finish] is called, the data will be persisted
+    /// in a single database transaction.
+    pub async fn new_transaction(&self) -> CentralContext {
+        CentralContext::new(self).await
+    }
+}
+
+impl CentralContext {
+    async fn new(central: &MlsCentral) -> Self {
+        let transaction = central.mls_backend.new_transaction();
+        let mls_groups = Arc::new(RwLock::new(Default::default()));
+        let callbacks = central.callbacks.clone();
+        let mls_client = central.mls_client.clone();
+        let lock_guard = Arc::new(central.transaction_lock.lock_arc().await);
+        Self {
+            state: Arc::new(
+                ContextState::Valid {
+                    mls_client,
+                    callbacks,
+                    transaction,
+                    mls_groups,
+                }
+                .into(),
+            ),
+            _lock_guard: lock_guard,
+        }
+    }
+
+    pub(crate) async fn mls_client(&self) -> CryptoResult<RwLockReadGuardArc<Option<Client>>> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client,
+                callbacks: _,
+                transaction: _,
+                mls_groups: _,
+            } => Ok(mls_client.read_arc().await),
```

Comment:
We can use `..` for unused fields
```suggestion
            ContextState::Valid { mls_client, .. } => Ok(mls_client.read_arc().await),
```